File: 백준/Silver/20436. ZOAC 3/ZOAC 3.cc
#include <iostream>
#include <map>
#include <cmath>
using namespace std;

struct MyPoint
{
    int x, y;
};

int main() {
    map<char, MyPoint> left;
    map<char, MyPoint> right;
    left['q'] = { 0,0 };
    left['w'] = { 0,1 };
    left['e'] = { 0,2 };
    left['r'] = { 0,3 };
    left['t'] = { 0,4 };
    left['a'] = { 1,0 };
    left['s'] = { 1,1 };
    left['d'] = { 1,2 };
    left['f'] = { 1,3 };
    left['g'] = { 1,4 };
    left['z'] = { 2,0 };
    left['x'] = { 2,1 };
    left['c'] = { 2,2 };
    left['v'] = { 2,3 };

    right['y'] = { 0,1 };
    right['u'] = { 0,2 };
    right['i'] = { 0,3 };
    right['o'] = { 0,4 };
    right['p'] = { 0,5 };
    right['h'] = { 1,1 };
    right['j'] = { 1,2 };
    right['k'] = { 1,3 };
    right['l'] = { 1,4 };
    right['b'] = { 2,0 };
    right['n'] = { 2,1 };
    right['m'] = { 2,2 };

    char l, r;
    cin >> l >> r;
    MyPoint currentLeft = left[l];
    MyPoint currentRight = right[r];

    string str;
    cin >> str;

    int count = 0;
    for (int i = 0; i < str.length(); i++)
    {
        count++;
        if (left.find(str[i]) != left.end())
        {
            count += abs(currentLeft.x - left[str[i]].x) + abs(currentLeft.y - left[str[i]].y);
            currentLeft = left[str[i]];
        }
        else if (right.find(str[i]) != right.end())
        {
            count += abs(currentRight.x - right[str[i]].x) + abs(currentRight.y - right[str[i]].y);
            currentRight = right[str[i]];
        }
    }
    cout << count;
    return 0;
}

File: 백준/Silver/20436. ZOAC 3/ZOAC 3.js
const fs = require('fs');
const filepath = process.platform === 'linux' ? '/dev/stdin' : './input.txt';
const input = fs.readFileSync(filepath).toString().trim().split('\n');

const [leftStart, rightStart] = input[0].trim().split(' ');
const str = input[1];

const left = new Map([
    ['q', { x: 0, y: 0 }], ['w', { x: 0, y: 1 }], ['e', { x: 0, y: 2 }], ['r', { x: 0, y: 3 }], ['t', { x: 0, y: 4 }],
    ['a', { x: 1, y: 0 }], ['s', { x: 1, y: 1 }], ['d', { x: 1, y: 2 }], ['f', { x: 1, y: 3 }], ['g', { x: 1, y: 4 }],
    ['z', { x: 2, y: 0 }], ['x', { x: 2, y: 1 }], ['c', { x: 2, y: 2 }], ['v', { x: 2, y: 3 }]
]); 

const right = new Map([
    ['y', { x: 0, y: 1 }], ['u', { x: 0, y: 2 }], ['i', { x: 0, y: 3 }], ['o', { x: 0, y: 4 }], ['p', { x: 0, y: 5 }],
    ['h', { x: 1, y: 1 }], ['j', { x: 1, y: 2 }], ['k', { x: 1, y: 3 }], ['l', { x: 1, y: 4 }],
    ['b', { x: 2, y: 0 }], ['n', { x: 2, y: 1 }], ['m', { x: 2, y: 2 }]
]);

let currentLeft = left.get(leftStart);
let currentRight = right.get(rightStart);

let count = 0;
for (const char of str) {
    count++;
    if (left.has(char)) {
        const newLeft = left.get(char);
        count += Math.abs(currentLeft.x - newLeft.x) + Math.abs(currentLeft.y - newLeft.y);
        currentLeft = newLeft;
    } else if (right.has(char)) {
        const newRight = right.get(char);
        count += Math.abs(currentRight.x - newRight.x) + Math.abs(currentRight.y - newRight.y);
        currentRight = newRight;
    }
}

console.log(count);
